sdk/include/opentelemetry/sdk/trace/sampler.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include "opentelemetry/common/attribute_value.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/trace/trace_id.h"
+#include "opentelemetry/version.h"
+
+#include <map>
+#include <memory>
+#include <string>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+
+/**
+ * A sampling Decision for a Span to be created.  
+ */
+enum class Decision
+{
+  // IsRecording() == false, span will not be recorded and all events and attributes will be
+  // dropped.
+  NOT_RECORD,
+  // IsRecording() == true, but Sampled flag MUST NOT be set.
+  RECORD,
+  // IsRecording() == true AND Sampled flag` MUST be set.
+  RECORD_AND_SAMPLE
+};
+
+/**
+ * The output of ShouldSample. 
+ * It contains a sampling Decision and a set of Span Attributes.
+ */
+struct SamplingResult
+{
+  Decision decision;
+  // A set of span Attributes that will also be added to the Span
```

Comment:
Can we add a note that this can be a `nullptr`? Most samplers will not return attributes, and in this way we can save quite a bit of heap allocations/stack space.